kernel: some cleanup; unify user interrupt handling between GAP and HPC-GAP

fingolfin opened the pull request:
Some assorted cleanup.

Major change is the addition of `FIRST_NON_INTERRUPT_STAT` and `LAST_NON_INTERRUPT_STAT` and the comment preceding them, which tries to rationalize the (to me) previously completely unmotivated and confusing way `InterruptExecStat` treated some stat types differently than others.

I wonder if @stevelinton can confirm or else correct this guess?

Also note that with that PR, I actually changed how user interrupts affect `T_RETURN_VOID`. It felt more logical that way, and I wonder if the original code treated it differently for actual reasons, or just by accident? Also, I think in reality it will virtually never make a user visible difference, but perhaps I am missing something?

We might also want to consider moving `T_EMPTY` and/or `T_ATOMIC` into that `*_NON_INTERRUPT_STAT` range... Not sure.

ChrisJefferson commented:
This seems good to me. I've never noticed GAP being slow to respond to ctrl+c, apart from when it is in C code in the kernel or a package, at which point the statements we break on won't make much difference anyway, so I would probably resist the urge to add more statements to the break set -- for example I can't imagine how we'd tell users we stopped at a `T_EMPTY`, as the whole point is it's an empty statement.

fingolfin commented:
@ChrisJefferson but it's the other way around: we currently *are* breaking on `T_EMPTY` (etc.), and I suggested that we should *not* do that.

ChrisJefferson commented:
Woops, I misread. I also agree we shouldn't be breaking on `T_EMPTY`.

fingolfin commented:
That is, we currently break on all types of statements, except for those in between `FIRST_NON_INTERRUPT_STAT` and `LAST_NON_INTERRUPT_STAT`.

Most of the statements in this set of "non-break" statements are what I'd call "compound statements", i.e., loops, if/else, and `T_SEQ*` -- the only compound statement type not in that range is `T_ATOMIC`, and I think the only reason for that is that it was added years after the other statement types, and I am sure people simply did not realize the significance of where in the list of statement types it was put (not surprising, given that the fact that it matters was rather hidden; part of this PR is of course about making it much easier to notice it).

Hence, I think we should move `T_ATOMIC` up, and e.g. put it right after `T_SEQ_STAT7`. I'd then also add a `FIRST_COMPOUND_STAT`/`LAST_COMPOUND_STAT` range around them (feel free to suggest a better name).

The other statements in the "non-break" range affect control flow: `continue`, `break`, `return`. Guess it kind of make sense to skip them, as they effectively take no time -- well, with the possible exception of `T_RETURN_OBJ`, which of course evaluates an expression... So I am somewhat puzzled as to why before this PR, we did not break on ``T_RETURN_VOID` but did break on `T_RETURN_OBJ`. My guess still is that this different treatment is due to a mistake, but I am just guessing.

fingolfin commented:
@ChrisJefferson I just updated this PR to add more ranges to the statement enum, and also moved `T_ATOMIC` and `T_EMPTY` to IMHO more appropriate places.
